ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java

```diff
@@ -319,14 +321,19 @@ public void rollback() throws org.omg.CORBA.SystemException,
 						&& (e1.errorCode < XAException.XA_RBEND))
 				    {
 					/*
-					 * Has been marked as rollback-only. We still
-					 * need to call rollback.
+					 * Has been marked as rollback-only. If the branch has not been
+					 * already rolled back (i.e. _rolledBack == true ), we still need
+					 * to call rollback.
 					 */
 				    } else if ((e1.errorCode == XAException.XAER_RMERR) || (e1.errorCode == XAException.XAER_RMFAIL)){
 				    	    try {
-				    	    	    _theXAResource.rollback(_tranID);
+							if (!this._rolledBack)
+							{
+				    	    	_theXAResource.rollback(_tranID);
+								this._rolledBack = true;
+							}
 				    	    } catch (XAException e2)
-				    	    {	
```

Comment:
@jmfinelli a bit of code formatting is here too, please can I ask you to move it out of this specific commit?

Reply:
Solved